rules/directory/glob.bzl

```diff
@@ -0,0 +1,87 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
```

Comment:
I wish to justify the cost of implementation and mainteinance of a second implementation of glob, but this time Starlark.

So this Starlark glob, allows you to define a directory in one package, and then run glob over it in different package. Are there other differences?

Using using a regular filegroup and a regular glob, you could just merge srcs from the directory rule, with the data from glob rules, right?

Reply:
This is the use case I have for it.

With this mechanism:
```
# My module extension
http_archive(name = "amd64_sysroot", url=".../amd64_sysroot.tar.gz")
http_archive(name = "arm64_sysroot", url=".../arm64_sysroot.tar.gz", build_file="BUILD.sysroot.bazel")
```

```
# BUILD.sysroot.bazel
directory(name = "sysroot", srcs = glob(["**/*"], exclude=["BUILD.bazel"])
```

```
# toolchain/BUILD.bazel

alias(
    name = "sysroot",
    actual = select({
        "@platforms//arch:amd64": "@amd64_sysroot//:sysroot",
        "@platforms//arch:arm64": "@arm64_sysroot//:sysroot"
        # Other sysroots (eg. RISK-V)
    })
)

subdirectory(
    name = "include_dir",
    directory = ":sysroot",
    path = "usr/include",
)

directory_glob(
    name = "shared_libraries",
    directory = ":sysroot",
    srcs = ["lib/*.so", "lib64/*.so"]
)

cc_sysroot_tool(
    name = "clang",
    directory = ":sysroot",
    executable = "usr/bin/clang",
)

cc_action_type_config(
    name = "compile_config",
    tool = ":clang",
    data = [":include_dir"],
)

cc_action_type_config(
    name = "link_config",
    tool = ":linker",
    data = [":shared_libraries"],
)

cc_toolchain(
    sysroot = ":sysroot",
    include_directories = [":include_dir"],
    action_type_configs = [":compile_config", ":link_config"]
)
```

Without it, toolchains would be significantly harder to define and understand, since every time you want to use anything related to toolchains, you would need to `select` over all possible sysroots.

For example, our rust toolchain would have to write:
```
rust_toolchain(
    ...,
    rustc = select({
        "@platforms//arch:amd64": "@amd64_sysroot//:rustc",
        ...
    })
```

This has two primary downsides:
1) If you ever decide you want to support another platform, with my design, I only have to update two lines (add the sysroot http_archive, and a condition to the select). Without it, I would have to update my whole codebase.
2) (this one is more minor) The top-level build file in `@*_sysroot` gets very massive without it, and there's not an easy way to organize it. Almost all this code would need to go into `BUILD.sysroot.bazel`, while in the current design I can split it up. Of course, you can somewhat mitigate this by providing extra build files, but then it won't work with http_archive.

Reply:
You could put the whole content of ` toolchain/BUILD.bazel` into `BUILD.sysroot.bazel` and use `filegroup` rules like:

```
filegroup(
    name = "include_dir",
    srcs = glob(["usr/include/**"]),
)

filegroup(
    name = "shared_libraries",
    srcs = glob(["lib/*.so", "lib64/*.so"]),
)

cc_sysroot_tool(
    name = "clang",
    executable = "usr/bin/clang",
)

cc_action_type_config(
    name = "compile_config",
    tool = ":clang",
    data = [":include_dir"],
)

cc_action_type_config(
    name = "link_config",
    tool = ":linker",
    data = [":shared_libraries"],
)

cc_toolchain(
    include_directories = [":include_dir"],
    action_type_configs = [":compile_config", ":link_config"]
)
```

Then you can do the selection based on the `cc_toolchain` if need be, or better yet use toolchain resolution.

I'm concerned about out of package globs, that used to be possible with `Fileset`s. Your implementation is a lot better though, because you're only projecting/selecting specific files from a large collection. It might be prone to more rebuilding (change one file and all the targets could be affected), but changes in prebuilt binary tools are probably unlikely.

The main benefit that I see is simplicity - you're using standard rules, and there's no maintenance costs associated with a new rule set. You could also address differences in directory structures among different platforms without introducing new selects.

Reply:
If this is going to be contentious, I'll split this into two PRs and we can can continue discussions on globbing in the other PR, since this is a hard blocker for the rules_cc work.